saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/infer/ConstrainedClassifier.scala

```diff
@@ -0,0 +1,336 @@
+    */
+  protected def pathToHead: Option[Edge[T, HEAD]] = None
+
+  private def deriveTestInstances: Iterable[T] = {
+    pathToHead.map(edge => edge.from)
+      .orElse({
+        onClassifier match {
+          case clf: Learnable[T] => Some(clf.node)
+          case _ => logger.error("pathToHead is not provided and the onClassifier is not a Learnable!"); None
+        }
+      })
+      .map(node => node.getTestingInstances)
+      .getOrElse(Iterable.empty)
+  }
+
+  def getCandidates(head: HEAD): Seq[T] = {
+    if (tType.equals(headType) || pathToHead.isEmpty) {
+      Seq(head.asInstanceOf[T])
+    } else {
+      val l = pathToHead.get.backward.neighborsOf(head)
+      l.size match {
+        case 0 =>
+          logger.error("Failed to find part")
+          Seq.empty[T]
+        case _ => l.filter(filter(_, head)).toSeq
+      }
+    }
+  }
+
+  private def findHead(x: T): Option[HEAD] = {
+    if (tType.equals(headType) || pathToHead.isEmpty) {
+      Some(x.asInstanceOf[HEAD])
+    } else {
+      val l = pathToHead.get.forward.neighborsOf(x).toSet.toSeq
+      l.length match {
+        case 0 =>
+          logger.trace("Failed to find head")
+          None
+        case 1 =>
+          logger.trace(s"Found head ${l.head} for child $x")
+          Some(l.head)
+        case _ =>
+          logger.trace("Found too many heads; this is usually because some instances belong to multiple 'head's")
+          Some(l.head)
+      }
+    }
+  }
+
+  /** given a solver type, instantiates a solver, uppon calling it */
+  private def getSolverInstance: ILPSolver = solverType match {
+    case OJAlgo => new OJalgoHook()
+    case Gurobi => new GurobiHook()
+    case Balas => new BalasHook()
+    case _ => throw new Exception("Hook not found! ")
+  }
+
+  /** given an instance */
+  def apply(t: T): String = {
+    findHead(t) match {
+      case Some(head) => build(head, t)
+      case None => onClassifier.classifier.discreteValue(t)
+    }
+  }
+
+  private def getInstancesInvolvedInProblem(constraintsOpt: Option[Constraint[_]]): Option[Set[_]] = {
+    constraintsOpt.map { constraint => getInstancesInvolved(constraint) }
+  }
+
+  private def getClassifiersInvolvedInProblem(constraintsOpt: Option[Constraint[_]]): Option[Set[LBJLearnerEquivalent]] = {
+    constraintsOpt.map { constraint => getClassifiersInvolved(constraint) }
+  }
+
+  /** given a head instance, produces a constraint based off of it */
+  private def instantiateConstraintGivenInstance(head: HEAD): Option[Constraint[_]] = {
+    // look at only the first level; if it is PerInstanceConstraint, replace it.
+    subjectTo.map {
+      case constraint: PerInstanceConstraint[HEAD] => constraint.sensor(head)
+      case constraint: Constraint[_] => constraint
+    }
+  }
+
+  /** find all the instances used in the definiton of the constraint. This is used in caching the results of inference  */
+  private def getInstancesInvolved(constraint: Constraint[_]): Set[_] = {
```

Comment:
The `getInstancesInvolved` and `getClassifiersInvolved` methods do not depend on ConstrainedClassifer (They don't need to be here). We should add these methods to the trait/abstract class for Constraint and have them implemented there.